### PR TITLE
[Nala]: Migrate to new ColorProvider

### PR DESCRIPTION
**Pull request by fallaciousreasoning**
Resolves https://github.com/brave/brave-browser/issues/40490

This will let us support theming in Nala C++ and make it much easier to use Nala tokens in views.

Related PR in Nala:
https://github.com/brave/leo/pull/775

## Submitter Checklist:

- [ ] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [ ] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [ ] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

**Comment by fallaciousreasoning** [comment]
@sangwoo108 @simonhong what are your thoughts on this?

File: ui/color/missing_color_ref_mixer.cc [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+  mixer[kColorRefTertiary35] = {palette->tertiary().get(35)};
+  mixer[kColorRefTertiary98] = {palette->tertiary().get(98)};
+}
+}  // namespace
+
+void AddMissingRefColorMixer(ColorProvider* provider,
+                             const ColorProviderKey& key) {
+  // Note: The logic for determining whether to use the baseline or generated
+  // color is taken from AddRefMixer in
+  // /home/jay/dev/brave-browser/src/ui/color/ref_color_mixer.cc
```
Comment (sangwoo108) [inline ui/color/missing_color_ref_mixer.cc:66]: nit:
```suggestion
  // //ui/color/ref_color_mixer.cc
```

File: ui/color/missing_color_ref_mixer.h [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+// You can obtain one at https://mozilla.org/MPL/2.0/.
+
+#ifndef BRAVE_UI_COLOR_MISSING_COLOR_REF_MIXER_H_
+#define BRAVE_UI_COLOR_MISSING_COLOR_REF_MIXER_H_
+
+#include "ui/color/color_provider.h"
+#include "ui/color/color_provider_key.h"
+
+namespace ui {
+void AddMissingRefColorMixer(ColorProvider* provider,
```
Comment (sangwoo108) [inline ui/color/missing_color_ref_mixer.h:13]: Q. I'm curious what would happen if we didn't have this.

File: patches/ui-color-color_id.h.patch [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+--- a/ui/color/color_id.h
++++ b/ui/color/color_id.h
+@@ -2,8 +2,8 @@
+ // Use of this source code is governed by a BSD-style license that can be
+ // found in the LICENSE file.
+ 
+-#ifndef UI_COLOR_COLOR_ID_H_
+-#define UI_COLOR_COLOR_ID_H_
++#ifndef COLOR_ID_H_
++#define COLOR_ID_H_
```
Comment (sangwoo108) [inline patches/ui-color-color_id.h.patch:12]: Can we revert this change?

File: browser/ui/views/overlay/brave_video_overlay_window_views.cc [hunk 4]
```diff
@@ -14,7 +14,7 @@
 #include "base/strings/utf_string_conversions.h"
 #include "brave/browser/ui/views/overlay/brave_back_to_tab_label_button.h"
 #include "brave/components/vector_icons/vector_icons.h"
-#include "brave/ui/color/leo/colors.h"
+#include "brave/node_modules/@brave/leo/tokens/skia/nala_color_id.h"
```
Comment (sangwoo108) [inline browser/ui/views/overlay/brave_video_overlay_window_views.cc:17]: Can we use the header copied into out/Foo/gen/... folder?

File: ui/color/missing_color_ref_mixer.h (hunk 2 above)
Comment (sangwoo108) [reply ui/color/missing_color_ref_mixer.h:13]: Is this something related to Material design's tokens we talked?

**Review by sangwoo108 — COMMENTED** [review]
@fallaciousreasoning Awesome work! I believe this would be a huge progress for UI dev folks. Love the idea that we should use nala color id and token without color scheme.

I'm not sure if it'd be good but even it would be possible to remove other color mixers than Nala color mixer, and use nala color id directly from each UI.

File: ui/color/missing_color_ref_mixer.h (hunk 2 above)
Comment (fallaciousreasoning) [reply ui/color/missing_color_ref_mixer.h:13]: Yeah, so basically the new Nala tokens are based on the dynamic material theme color. Unfortunately, they don't add all the tones to the color mixer, so we have to add the missing ones we depend on :laughing:

**Comment by fallaciousreasoning** [comment]
@sangwoo108 we probably can't remove all the other color mixers - the Nala theme is actually generated from the Chromium Material UI color refs and we probably want to make sure all the Chromium ones stick around (to make rebasing easier).

File: chromium_src/check_chromium_src_config.json5 [hunk 5]
```diff
@@ -116,6 +116,9 @@
             "TargetGetModuleFileNameExA64",
             "TargetGetModuleFileNameExW64",
         ],
+        "ui/color/color_id.h": [
+            "MATERIAL_MISSING_COLOR_IDS"
+        ],
```
Comment (sangwoo108) [inline chromium_src/check_chromium_src_config.json5:121]: Q. Is this for ignoring warning saying something like "no symbol was found in the associated files?"

File: chromium_src/ui/color/color_id.h [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+  E_CPONLY(kColorRefSecondary98)   \
+  E_CPONLY(kColorRefTertiary5)     \
+  E_CPONLY(kColorRefTertiary15)    \
+  E_CPONLY(kColorRefTertiary25)    \
+  E_CPONLY(kColorRefTertiary35)    \
+  E_CPONLY(kColorRefTertiary98)
+
+#include "src/ui/color/color_id.h"  // IWYU pragma: export
+
+#endif  // BRAVE_CHROMIUM_SRC_UI_COLOR_COLOR_ID_H_
```
Comment (sangwoo108) [inline chromium_src/ui/color/color_id.h:30]: Should we undef MATERIAL_MISSING_COLOR_IDS?

**Review by sangwoo108 — APPROVED** [review]
LGTM 👍 Excellent work!

File: ui/color/missing_color_ref_mixer.cc [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+#include "brave/ui/color/missing_color_ref_mixer.h"
+
+#include "ui/color/color_provider.h"
+#include "ui/color/color_recipe.h"
+#include "ui/color/dynamic_color/palette.h"
+#include "ui/color/dynamic_color/palette_factory.h"
+
+namespace ui {
+namespace {
+void AddBaseLine(ColorProvider* provider) {
```
Comment (sangwoo108) [inline ui/color/missing_color_ref_mixer.cc:15]: optional:  AddStatic()/AddDynamic() might be better by the definition of m3 material's color scheme, if I understand things correctly https://m3.material.io/styles/color/choosing-a-scheme

File: chromium_src/ui/color/color_id.h (hunk 6 above)
Comment (fallaciousreasoning) [reply chromium_src/ui/color/color_id.h:30]: The `COLOR_IDS` macro is used in a few places, so this needs to still be defined so that that expansion still works

File: chromium_src/check_chromium_src_config.json5 (hunk 5 above)
Comment (fallaciousreasoning) [reply chromium_src/check_chromium_src_config.json5:121]: Nah, its for not triggering PRESUBMIT because I don't undefine the macro again at the end of the file

Comment (sangwoo108) [reply chromium_src/check_chromium_src_config.json5:121]: Aha! Thanks

File: ui/color/missing_color_ref_mixer.cc (hunk 7 above)
Comment (fallaciousreasoning) [reply ui/color/missing_color_ref_mixer.cc:15]: Oh just remember why I did this - it's the same names as the `ref_color_mixer` in Chromium uses

Comment (sangwoo108) [reply ui/color/missing_color_ref_mixer.cc:15]: Aha, I see. Never mind then

File: browser/ui/color/brave_color_mixer.cc [hunk 8]
```diff
@@ -475,15 +447,12 @@ void AddBraveLightThemeColorMixer(ui::ColorProvider* provider,
   mixer[kColorBookmarkBarInstructionsText] = {
       PickColorContrastingToToolbar(key, mixer, SkColorSetRGB(0x49, 0x50, 0x57),
                                     SkColorSetRGB(0xFF, 0xFF, 0xFF))};
-  mixer[kColorBookmarkBarInstructionsLink] = {PickColorContrastingToToolbar(
-      key, mixer, leo::light::kColorTextInteractive,
-      leo::dark::kColorTextInteractive)};
+  mixer[kColorBookmarkBarInstructionsLink] = {nala::kColorTextInteractive};
```
Comment (simonhong) [inline browser/ui/color/brave_color_mixer.cc:450]: As we use same nala tokens in different themes,
I think most of them could be de-dup in `AddBraveColorMixerForAllThemes()`. It could be done as f/u.
Maybe we could use only one mixer `AddBraveColorMixerForAllThemes()` instead of using different mixers per themes after we set nala tokens for all color ids. :)

**Review by simonhong — APPROVED** [review]
++ 👍🏼

File: patches/ui-color-color_id.h.patch [hunk 9]
```diff
@@ -0,0 +1,13 @@
+diff --git a/ui/color/color_id.h b/ui/color/color_id.h
+index c9df38346b84c..7bb58bb01a476 100644
+--- a/ui/color/color_id.h
++++ b/ui/color/color_id.h
+@@ -674,7 +674,7 @@
+ 
+ #define COLOR_IDS \
+   CROSS_PLATFORM_COLOR_IDS \
+-  PLATFORM_SPECIFIC_COLOR_IDS
++  PLATFORM_SPECIFIC_COLOR_IDS MATERIAL_MISSING_COLOR_IDS
```
Comment (goodov) [inline patches/ui-color-color_id.h.patch:10]: can't you insert `MATERIAL_MISSING_COLOR_IDS` between? like so:
```
   CROSS_PLATFORM_COLOR_IDS \
+  MATERIAL_MISSING_COLOR_IDS \
   PLATFORM_SPECIFIC_COLOR_IDS
```

also it should be named `BRAVE_...`

I believe this way you don't need to "fix" chromium_src checker.

File: package.json [hunk 10]
```diff
@@ -174,7 +174,7 @@
   },
   "dependencies": {
     "@brave/brave-ui": "0.40.6",
-    "@brave/leo": "github:brave/leo#63e98a1bb14cf87dcb8cbd4436e9e38615840447",
+    "@brave/leo": "github:brave/leo#skia-colors",
```
Comment (goodov) [inline package.json:177]: this is not the final form, right? I expect to have a commit here.

File: chromium_src/ui/color/color_mixers.cc [hunk 11]
```diff
@@ -0,0 +1,17 @@
+// Copyright (c) 2024 The Brave Authors. All rights reserved.
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at https://mozilla.org/MPL/2.0/.
+
+#include "brave/ui/color//nala/nala_color_mixer.h"
+#include "brave/ui/color/missing_color_ref_mixer.h"
+#include "ui/color/ref_color_mixer.h"
+
+#define AddRefColorMixer                  \
```
Comment (goodov) [inline chromium_src/ui/color/color_mixers.cc:10]: `(...)` and use `(__VA_ARGS__)` on all lines?

File: chromium_src/ui/color/color_id.h [hunk 12]
```diff
@@ -0,0 +1,30 @@
...
+// You can obtain one at https://mozilla.org/MPL/2.0/.
+
+#ifndef BRAVE_CHROMIUM_SRC_UI_COLOR_COLOR_ID_H_
+#define BRAVE_CHROMIUM_SRC_UI_COLOR_COLOR_ID_H_
+
+// These are Material colors which Chromium doesn't add to its color mixer but
+// which we need for Nala. We have a mixer in
+// brave/ui/color/missing_color_ref_mixer.h for adding these missing Material
+// colors.
+#define MATERIAL_MISSING_COLOR_IDS \
```
Comment (goodov) [inline chromium_src/ui/color/color_id.h:13]: if this should always be defined, pls add a comment.

File: ui/color/BUILD.gn [hunk 13]
```diff
@@ -3,17 +3,39 @@
...
+  deps = [ ":copy_nala_color_ids" ]
+}
+
+copy("copy_nala_color_mixer") {
+  sources = [
+    "//brave/node_modules/@brave/leo/tokens/skia/nala_color_mixer.cc",
+    "//brave/node_modules/@brave/leo/tokens/skia/nala_color_mixer.h",
+  ]
+
+  outputs = [ "$root_out_dir/gen/brave/ui/color/nala/{{source_file_part}}" ]
```
Comment (goodov) [inline ui/color/BUILD.gn:24]: I don't think this is correct, because in cross-compile mode `gen` may be located one level deeper. Shouldn't this be `$target_gen_dir/{{source_file_part}}` instead? You may need to rename the target to `nala` for things to work.

same for the above target (not sure why these are separate targets, just merge them into one, copying is not cpu-intensive operation during the build, so it can just be a single target).

File: ui/color/missing_color_ref_mixer.h (hunk 2 above)
Comment (goodov) [inline ui/color/missing_color_ref_mixer.h:13]: why is it called `Missing`? I don't see any colors with "Missing" name in it. Can this have a better name? Maybe just "Brave" or "Nala" or something that make more sense long-term?

File: package.json (hunk 10 above)
Comment (fallaciousreasoning) [reply package.json:177]: Yup, I mentioned this in the PR description - I don't want to merge the `skia-colors` branch into Nala until we're in agreement about what shape this is going to be in `brave-core` because it'll block bug fixes.

File: ui/color/missing_color_ref_mixer.h (hunk 2 above)
Comment (fallaciousreasoning) [reply ui/color/missing_color_ref_mixer.h:13]: These are the colors from the Material palette which are missing from the `RefColorMixer` - essentially, they're generated tones from the Material UI palette (`RefColor`s in Chromium) which Nala requires.

Added a comment.

File: ui/color/BUILD.gn (hunk 13 above)
Comment (fallaciousreasoning) [reply ui/color/BUILD.gn:24]: Wow, didn't know about the `gen` folder being located in different places, but it makes sense! I've replaced it with `$root_gen_dir` instead, which I think should work :smile:

File: patches/ui-color-color_id.h.patch (hunk 9 above)
Comment (fallaciousreasoning) [reply patches/ui-color-color_id.h.patch:10]: Renamed & formatted - Unfortunately I still have to add the entry in `check_chromium_src_config.json5` for it to pass

File: ui/color/BUILD.gn (hunk 13 above)
Comment (goodov) [reply ui/color/BUILD.gn]: > Wow, didn't know about the `gen` folder being located in different places, but it makes sense! I've replaced it with `$root_gen_dir` instead, which I think should work, and merged the copy targets 😄

have you tried `$target_gen_dir`? GN should fill it with `gen/brave/ui/color/{target_name}` path for you.

**Review by thypon — APPROVED** [review]

**Review by goodov — APPROVED** [review]

File: chromium_src/ui/color/color_id.h [hunk 14]
```diff
@@ -0,0 +1,32 @@
...
+// You can obtain one at https://mozilla.org/MPL/2.0/.
+
+#ifndef BRAVE_CHROMIUM_SRC_UI_COLOR_COLOR_ID_H_
+#define BRAVE_CHROMIUM_SRC_UI_COLOR_COLOR_ID_H_
+
+// These are Material colors which Chromium doesn't add to its color mixer but
+// which we need for Nala. We have a mixer in
+// brave/ui/color/missing_color_ref_mixer.h for adding these missing Material
+// colors.
+// This isn't defined because other files use the |COLOR_ID| macro which depends
```
[the remark this answers is not in the document]
Comment (fallaciousreasoning) [reply chromium_src/ui/color/color_id.h:13]: :facepalm:

**Comment by fallaciousreasoning** [comment]
Okay, I've updated this PR to point to the commit sha now we've landed the PR in Nala
